hw2/hw2.py

```diff
@@ -0,0 +1,420 @@
+
+        This function has no return value
+        """
+        if self.depth >= self.max_depth or len(self.data) == 0:
+            self.terminal = True
+            return  # Stop splitting if max depth is reached or no data is present
+
+        # Select the best feature based on the impurity reduction
+        best_goodness = -np.inf
+        best_feature = None
+        best_splits = {}
+
+        for feature in range(self.data.shape[1] - 1):  # Assume last column is the label
+            goodness, splits = self.goodness_of_split(feature, self.gain_ratio)
+            if goodness > best_goodness:
+                best_goodness = goodness
+                best_feature = feature
+                best_splits = splits
+
+        # Check if splitting improves or possible pruning conditions
+        if best_goodness <= 0 or not best_splits:
+            self.terminal = True
+            return  # No improvement in impurity or no splits available
+
+        # Chi square pruning
+        if self.chi < 1:
+            # Chi square test
+            chi_square_val = 0
+            degree_of_freedom = (
+                1  # TODO - calc degree? # TODO(#attributes-1)(#classes-1)
+            )
+            # TODO - is the degree_of_freedom dynamic ? meaning - #attributes-1 -depth
+
+            labels = self.data[:, -1]
+            unique_labels, counts = np.unique(labels, return_counts=True)
+
+            # sum over selected feature values
+            for _, subset in best_splits.items():
+                sub_labels = subset[:, -1]
+
+                for idx, label in enumerate(unique_labels):
+                    expected = len(subset) * (counts[idx] / counts.sum())
+                    actual_sub_indices = np.where(sub_labels == label)[0]
+                    chi_square_val += (
+                        (len(actual_sub_indices) - expected) ** 2
+                    ) / expected
+
+            if chi_square_val <= chi_table[degree_of_freedom][self.chi]:
+                # Split is "random", pruning
+                self.terminal = True
+                return
+
+        # Create child nodes for each split
+        self.feature = best_feature
+        for val, subset in best_splits.items():
+            if len(subset) > 0:  # Only create a child if there are data points
+                child_node = DecisionNode(
+                    data=subset,
+                    impurity_func=self.impurity_func,
+                    depth=self.depth + 1,
+                    chi=self.chi,
+                    max_depth=self.max_depth,
+                    gain_ratio=self.gain_ratio,
+                )
+                self.add_child(child_node, val)
+
+
+class DecisionTree:
+    def __init__(
+        self, data, impurity_func, feature=-1, chi=1, max_depth=1000, gain_ratio=False
+    ):
+        self.data = data  # the relevant data for the tree
+        self.impurity_func = (
```

Comment:
Is that a result of a linter ?
*Change in a predefined code.

Reply:
Linter, doesn't change anything